en/tutorials-and-examples/cms/authentication-plugin.rst

```diff
@@ -0,0 +1,288 @@
+    {
+        $authenticationService = new \Authentication\AuthenticationService();
+
+        // Load identifiers, ensure we check email and password fields
+        $authenticationService->loadIdentifier('Authentication.Password', [
+            'fields' => [
+                'username' => 'email',
+                'password' => 'password',
+            ]
+        ]);
+
+        // Load the authenticators, you want session first
+        $authenticationService->loadAuthenticator('Authentication.Session');
+        // Configure form data check to pick email and password
+        $authenticationService->loadAuthenticator('Authentication.Form', [
+            'fields' => [
+                'username' => 'email',
+                'password' => 'password',
+            ],
+            'loginUrl' => '/users/login',
+        ]);
+
+        return $authenticationService;
+    }
+
+In you AppController class add the following code
+
+    // src/Controller/AppController.php
+    public function initialize(): void
+    {
+        parent::initialize();
+        $this->loadComponent('RequestHandler');
+        $this->loadComponent('Flash');
+        // add this like to check authentication result and lock your site
+        $this->loadComponent('Authentication.Authentication');
+
+After adding this code, on every request, the AuthenticationMiddleware will inspect
+the request session to look for an authenticated user, and if we are in /users/login
+page, it'll inspect also the posted form data (if any) to extract the credentials.
+By default the credentials will be extracted from the ``username`` and ``password``
+fields in the request data.
+The authentication result will be injected in a request attribute named
+``authentication``. You can inspect the result at any time using
+``$this->request->getAttribute('authentication')`` from your controller actions.
+All your pages will be locked, as the ``AuthenticationComponent`` is checking the
+result on every request, and not finding any authenticated user, it'll redirect the
+user to the ``/users/login`` page.
+Note at this point, the site won't work as we don't have a login page yet...
+If you visit your site, you'll get an "infinite redirect loop" so let's fix that...
+
+In your UsersController, add the following code
+
+    public function beforeFilter(\Cake\Event\EventInterface $event)
+    {
+        parent::beforeFilter($event);
+        // configure the login action to don't require authentication, preventing
+        // the infinite redirect loop issue
+        $this->Authentication->addUnauthenticatedActions(['login']);
+    }
+
+    public function login()
+        $this->request->allowMethod(['get', 'post']);
+        $result = $this->Authentication->getResult();
+        // regardless of POST or GET, redirect if user is logged in
+        if ($result->isValid()) {
+            // redirect to /pages/home after login success
+            $redirect = $this->request->getQuery('redirect', [
```

Comment:
Could we use `$this->Authentication->getLoginRedirect() ?? ['controller' => 'Pages', 'action' => 'display', 'home']` instead? This approach leaves the application with an open-redirect.